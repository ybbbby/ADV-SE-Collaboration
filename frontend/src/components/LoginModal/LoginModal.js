import React from 'react'
import PropTypes from 'prop-types'
import {
  Button,
  Dialog,
  DialogContent,
  DialogTitle,
  IconButton,
  Typography,
  SvgIcon,
} from '@material-ui/core'
import { makeStyles } from '@material-ui/core/styles'
import CloseIcon from '@material-ui/icons/Close'
import login from '../../api/login'

const useStyles = makeStyles((theme) => ({
  closeButton: {
    position: 'absolute',
    right: theme.spacing(1),
    top: theme.spacing(1),
    color: theme.palette.grey[500],
  },
  content: {
    padding: theme.spacing(5.5),
    textAlign: 'center',
  },
  loginButton: {
    margin: theme.spacing(1),
    padding: theme.spacing(1.5),
  },
  google: {
    verticalAlign: 'middle',
    marginRight: theme.spacing(1),
  },
}))

function GoogleIcon(props) {
  return (
    <SvgIcon {...props}>
      <path d="M21.35,11.1H12.18V13.83H18.69C18.36,17.64 15.19,19.27 12.19,19.27C8.36,19.27 5,16.25 5,12C5,7.9 8.2,4.73 12.2,4.73C15.29,4.73 17.1,6.7 17.1,6.7L19,4.72C19,4.72 16.56,2 12.1,2C6.42,2 2.03,6.8 2.03,12C2.03,17.05 6.16,22 12.25,22C17.6,22 21.5,18.33 21.5,12.91C21.5,11.76 21.35,11.1 21.35,11.1V11.1Z" />
    </SvgIcon>
  )
}

function LoginModal(props) {
  const classes = useStyles()
  const { handleClose, open } = props

  function clickLogin() {
    login().then((data) => (window.location.href = data))
  }

  return (
    <Dialog
      open={open}
      onClose={() => handleClose()}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">
        <IconButton
          aria-label="close"
          className={classes.closeButton}
          onClick={() => handleClose()}
        >
          <CloseIcon />
        </IconButton>
      </DialogTitle>
      <DialogContent className={classes.content}>
        <Typography variant="h4" gutterBottom>
          Welcome Back!
        </Typography>
        <Typography variant="subtitle1" gutterBottom>
          Sign in to get personalized event recommendations based on your
          location. Explore, save and join the events you like, and interact
          with other users nearby.
        </Typography>
        <Button
          variant="outlined"
          className={classes.loginButton}
          onClick={clickLogin}
        >
          <GoogleIcon className={classes.google} />
          Sign in with Google
        </Button>
      </DialogContent>
    </Dialog>
  )
}

LoginModal.propTypes = {
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
}

export default LoginModal
