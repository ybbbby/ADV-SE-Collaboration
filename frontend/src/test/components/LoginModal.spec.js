/* eslint-env jest */
import { shallow } from 'enzyme'
import React from 'react'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import LoginModal from '../../components/LoginModal/LoginModal'

describe('LoginModal', () => {
  test('does not render modal when closed', () => {
    let openLogin = false
    const setOpenLogin = jest.fn(() => (openLogin = false))
    render(<LoginModal handleClose={() => setOpenLogin()} open={openLogin} />)
    expect(screen.queryByText('Welcome Back!')).toBeNull()
  })

  test('renders modal when opened', () => {
    let openLogin = true
    const setOpenLogin = jest.fn(() => (openLogin = false))
    render(<LoginModal handleClose={() => setOpenLogin()} open={openLogin} />)
    expect(screen.queryByText('Welcome Back!')).toBeDefined()
  })

  test('click sign in button', async () => {
    fetch.mockResponseOnce('/')
    let openLogin = true
    const setOpenLogin = jest.fn(() => (openLogin = false))
    render(<LoginModal handleClose={() => setOpenLogin()} open={openLogin} />)
    await waitFor(() =>
      userEvent.click(
        screen.getByRole('button', { name: 'Sign in with Google' })
      )
    )
    expect(screen.queryByText('Welcome Back!')).toBeDefined()
  })

  test('close modal', async () => {
    let openLogin = true
    const setOpenLogin = jest.fn(() => (openLogin = false))
    render(<LoginModal handleClose={() => setOpenLogin()} open={openLogin} />)
    await waitFor(() =>
      userEvent.click(screen.getByRole('button', { name: 'close' }))
    )
    expect(setOpenLogin).toHaveBeenCalledTimes(1)
  })

  test('close modal by clicking away', () => {
    let openLogin = true
    const setOpenLogin = jest.fn(() => (openLogin = false))
    render(<LoginModal handleClose={() => setOpenLogin()} open={openLogin} />)
    fireEvent.keyDown(screen.getByRole('presentation'), {
      key: 'Escape',
      code: 'Escape',
    })
    expect(setOpenLogin).toHaveBeenCalledTimes(1)
  })

  it('snapshot when open', () => {
    let openLogin = true
    const setOpenLogin = jest.fn(() => (openLogin = false))
    const wrapper = shallow(
      <LoginModal handleClose={() => setOpenLogin()} open={openLogin} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('snapshot when closed', () => {
    let openLogin = false
    const setOpenLogin = jest.fn(() => (openLogin = false))
    const wrapper = shallow(
      <LoginModal handleClose={() => setOpenLogin()} open={openLogin} />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
