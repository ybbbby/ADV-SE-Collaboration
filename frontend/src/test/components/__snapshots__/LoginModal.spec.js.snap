// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginModal snapshot when closed 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="form-dialog-title"
  onClose={[Function]}
  open={false}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="form-dialog-title"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="close"
      className="makeStyles-closeButton-21"
      onClick={[Function]}
    >
      <Memo([object Object]) />
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className="makeStyles-content-22"
  >
    <WithStyles(ForwardRef(Typography))
      gutterBottom={true}
      variant="h4"
    >
      Welcome Back!
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      gutterBottom={true}
      variant="subtitle1"
    >
      Sign in to get personalized event recommendations based on your location. Explore, save and join the events you like, and interact with other users nearby.
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Button))
      className="makeStyles-loginButton-23"
      onClick={[Function]}
      variant="outlined"
    >
      <GoogleIcon
        className="makeStyles-google-24"
      />
      Sign in with Google
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogContent))>
</WithStyles(ForwardRef(Dialog))>
`;

exports[`LoginModal snapshot when open 1`] = `
<WithStyles(ForwardRef(Dialog))
  aria-labelledby="form-dialog-title"
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(DialogTitle))
    id="form-dialog-title"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-label="close"
      className="makeStyles-closeButton-21"
      onClick={[Function]}
    >
      <Memo([object Object]) />
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(DialogTitle))>
  <WithStyles(ForwardRef(DialogContent))
    className="makeStyles-content-22"
  >
    <WithStyles(ForwardRef(Typography))
      gutterBottom={true}
      variant="h4"
    >
      Welcome Back!
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Typography))
      gutterBottom={true}
      variant="subtitle1"
    >
      Sign in to get personalized event recommendations based on your location. Explore, save and join the events you like, and interact with other users nearby.
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(Button))
      className="makeStyles-loginButton-23"
      onClick={[Function]}
      variant="outlined"
    >
      <GoogleIcon
        className="makeStyles-google-24"
      />
      Sign in with Google
    </WithStyles(ForwardRef(Button))>
  </WithStyles(ForwardRef(DialogContent))>
</WithStyles(ForwardRef(Dialog))>
`;
